### PR TITLE
hbone: initial echo server/client implementation

### PR DESCRIPTION
**Pull request by howardjohn**
Per https://docs.google.com/document/d/1Ofqtxqzk-c_wn0EgAXjaJXDHB9KhDuLe-W3YGG67Y8g/edit

### CODE REVIEW

**Comment by costinm** [comment]
I would really appreciate it if we could have this in istio/pkg repo
instead of istio/istio.

Preferably with a separate go.mod

File: pkg/config/protocol/instance.go [hunk 1]
```diff
@@ -54,6 +54,8 @@ const (
 	Redis Instance = "Redis"
 	// MySQL declares that the port carries MySQL traffic.
 	MySQL Instance = "MySQL"
+	// HBONE declares that the port carries HBONE traffic.
+	HBONE Instance = "HBONE"
```
Comment (costinm) [inline pkg/config/protocol/instance.go:58]: Don't we have a dedicated port ?

Comment (costinm) [reply pkg/config/protocol/instance.go:58]: Or 2 ports actually - one for HBone and one for 'plain text/trusted network' hbone, i.e. h2 or h2c

File: pkg/hbone/README.md [hunk 2]
```diff
@@ -0,0 +1,71 @@
+# HTTP Based Overlay Network (HBONE)
+
+HTTP Based Overlay Network (HBONE) is the protocol used by Istio for communication between workloads in the mesh.
+At a high level, the protocol consists of tunneling TCP connections over HTTP/2 CONNECT, over mTLS.
```
Comment (costinm) [inline pkg/hbone/README.md:4]: I would skip 'over mTLS', and add 'over HTTP/2 and H3 CONNECT' ( not clear if QUIC is mTLS, or if in a trusted networks we must have outer mTLS or H2C is allowed )

File: pkg/hbone/README.md [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+## Specification
+
+TODO
```
Comment (costinm) [inline pkg/hbone/README.md:8]: My doc on gateway improvements has several links, you can add them here ( RFCs, BeyondCorp, browser specs)

File: pkg/config/protocol/instance.go (hunk 1 above)
Comment (howardjohn) [reply pkg/config/protocol/instance.go:58]: This is not something a user can configure by portname on Service -- but used for internal. Not by convention, this is enforced.

File: pkg/hbone/dialer.go [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+	// handle upstream (mtp server) <-- downstream (app)
+	copyBuffered(pw, conn, log.WithLabels("name", "conn to pipe"))
+
+	wg.Wait()
+	log.Info("stream closed in ", time.Since(t0))
+	return nil
+}
+
+// TLSDialWithDialer is an implementation of tls.DialWithDialer that accepts a generic Dialer
+func TLSDialWithDialer(dialer Dialer, network, addr string, config *tls.Config) (*tls.Conn, error) {
```
Comment (costinm) [inline pkg/hbone/dialer.go:139]: Why do we need this ? Normal pattern is create Dialer with config and call DialContext/Dial

File: pkg/hbone/dialer.go [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+	if network != "tcp" {
+		return net.Dial(network, address)
+	}
+	// TODO: use context
+	c, s := net.Pipe()
+	go func() {
+		err := proxyTo(s, d.cfg, address)
+		log.Infof("Tunnel complete: %v", err)
+	}()
+	return c, nil
```
Comment (costinm) [inline pkg/hbone/dialer.go:66]: I think we need to work on this - see my previous implementation, the HTTP RT should probably be done part of Dial, with proxy remaining in background. So we can return errors properly ( similar with how normal Dial works - i.e. a real connection/stream will be established by dial)

File: pkg/hbone/dialer.go (hunk 4 above)
Comment (howardjohn) [reply pkg/hbone/dialer.go:139]: I am not sure another way to add TLS to a Dialer?

File: pkg/hbone/dialer.go (hunk 5 above)
Comment (howardjohn) [reply pkg/hbone/dialer.go:66]: Agreed

**Comment by costinm** [comment]
On Mon, Jun 27, 2022 at 1:55 PM Nathan Mittler ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In pkg/config/protocol/instance.go
> <https://github.com/istio/istio/pull/39645#discussion_r907720073>:
>
> > @@ -54,6 +54,8 @@ const (
>  	Redis Instance = "Redis"
>  	// MySQL declares that the port carries MySQL traffic.
>  	MySQL Instance = "MySQL"
> +	// HBONE declares that the port carries HBONE traffic.
> +	HBONE Instance = "HBONE"
>
> Will HBONE always be over h2? If we ever supported h3, would it be
> considered a different protocol?
>
AFAIK we treat H2 and H3 as the same protocol - it really requires
'bidirectional stream support' which is missing in H1.

File: pkg/config/protocol/instance.go (hunk 1 above)
Comment (hzxuzhonghu) [inline pkg/config/protocol/instance.go:58]: Can we not add it here

Comment (howardjohn) [reply pkg/config/protocol/instance.go:58]: We would have to update all of our tests and echo codebase. Any concerns ? It cannot be used by real Service (which I plan to add a comment and test asserting this)

Comment (hzxuzhonghu) [reply pkg/config/protocol/instance.go:58]: My  concern is test code should not couple with main. Could be confused at a new protocol?

[the remark this answers is not in the document]
Comment (howardjohn) [reply pkg/config/protocol/instance.go:58]: In the future we may make it support h3. TBD if we consider that another protocol here.

This is just internal details here, not API, so we can change

File: pkg/hbone/dialer.go [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+	"golang.org/x/net/proxy"
+
+	"istio.io/istio/security/pkg/pki/util"
+	istiolog "istio.io/pkg/log"
+)
+
+var log = istiolog.RegisterScope("hbone", "", 0)
+
+type Config struct {
+	ProxyAddress string
```
[the remark this answers is not in the document]
Comment (howardjohn) [reply pkg/hbone/dialer.go:38]: putting it in comments, but ProxyAddress is the HBONE proxy, address is the destination address

File: pkg/hbone/server.go [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+	// Need to set this to allow timeout on the read header
+	h1 := &http.Transport{
+		ExpectContinueTimeout: 3 * time.Second,
+	}
+	h2, _ := http2.ConfigureTransports(h1)
+	h2.ReadIdleTimeout = 10 * time.Minute // TODO: much larger to support long-lived connections
+	h2.AllowHTTP = true
+	h2.StrictMaxConcurrentStreams = false
+	h2Server := &http2.Server{}
+	handler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
```
[the remark this answers is not in the document]
Comment (howardjohn) [reply pkg/hbone/server.go:38]: `r` and `w` are the standard names for this defined by go stdlib

File: pkg/hbone/util.go [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+	"net/http"
+	"sync"
+	"time"
+
+	istiolog "istio.io/pkg/log"
+)
+
+// createBuffer to get a buffer. io.Copy uses 32k.
+// experimental use shows ~20k max read with Firefox.
+var bufferPoolCopy = sync.Pool{New: func() interface{} {
```
[the remark this answers is not in the document]
Comment (howardjohn) [reply pkg/hbone/util.go:29]: Where are we locking? AFAIK sync.Pool uses no locking internally and I don't see any locking added here?

File: pkg/config/protocol/instance.go (hunk 1 above)
Comment (nmittler) [reply pkg/config/protocol/instance.go:59]: I think "experimental" is more accurate than "test". This only exists in this file. Seems that an adequate "proceed at your own risk" banner should be sufficient, no?

File: pkg/hbone/README.md (hunk 2 above)
Comment (nmittler) [reply pkg/hbone/README.md:4]: For now, isn't it mTLS? I'd rather keep the text representative of what has been built rather than what might be built in the future.

File: pkg/hbone/README.md (hunk 3 above)
Comment (nmittler) [reply pkg/hbone/README.md:8]: +1 ... could you fill out this section?

File: pkg/hbone/server.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (nmittler) [reply pkg/hbone/server.go:37]: doesn't make it clearer :). It's fine ... marking resolved.

File: pkg/hbone/util.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (nmittler) [reply pkg/hbone/util.go:29]: A couple years old, but [hare are some benchmarks](https://medium.com/swlh/go-the-idea-behind-sync-pool-32da5089df72) ... not sure if an extra couple of nanos matters much here :). Go doesn't implement thread-local storage anyway, so you can probably safely ignore my comment.

File: pkg/test/echo/server/forwarder/config.go [hunk 9]
```diff
@@ -55,13 +55,24 @@ type Config struct {
 	urlPath                 string
 	method                  string
 	secure                  bool
+
+	// completed whether the Config is complete or still in progress
+	completed bool
```
Comment (nmittler) [inline pkg/test/echo/server/forwarder/config.go:60]: why did you need this?

Comment (howardjohn) [reply pkg/test/echo/server/forwarder/config.go:60]: tbh I forget.. some issue about a race condition calling the setup twice. Does seem suspicious, let me look into it

**Review by nmittler — APPROVED** [review]

**Comment by howardjohn** [comment]
Figured it out, to newDialer we now pass config which is later mutated causing issues. IMO https://github.com/istio/istio/pull/39735 is a better fix (assuming it doesn't break anything, i split it out to make it more clear).

**Comment by costinm** [comment]
Can we separate the hbone client from the service port name ? I think it's
a very different discussion, better handled in API/doc.

**Comment by howardjohn** [comment]
This doesn't impact the service port - "HBONE" is still not a valid service
port, the protocol enum we are adding to is not implicitly adding it as a
valid option for services
